feat(heaphook) Define `AgnocastSharedMemoryAllocator` trait to support custom memory allocators

r1ru opened the pull request:
## Description
In the current implementation, we only support the TLSF allocator. As a first step toward supporting custom memory allocators in the future, I have defined a trait for memory allocators.

## Related links

## How was this PR tested?

- [ ] Autoware (required)
- [x] `bash scripts/e2e_test_1to1_with_ros2sub` (required)
- [x] `bash scripts/e2e_test_2to2` (required)
- [ ] kunit tests (required when modifying the kernel module)
- [ ] sample application

## Notes for reviewers
Although it may look like there are many changes, the only essential modification is the introduction of the trait. The TLSF-related definitions have been moved into tlsf.rs.

Koichi98 commented:
@r1ru 
Is the next step going to be an abstraction of the current direct calls to TLSF methods from lib.rs using dyn? I'd like to get a rough idea of the future PR plans.

r1ru commented:
@Koichi98 

Yes. Here is the code. I will make a few minor adjustments and then create a PR:
https://github.com/tier4/agnocast/tree/heaphook-custom-allocator-support

I renamed the `AgnocastSharedMemoryAllocator` trait to `SharedMemoryAllocator`,
and modified the code to use `AGNOCAST_SHARED_MEMORY_ALLOCATOR` instead of `TLSF`:

https://github.com/tier4/agnocast/blob/heaphook-custom-allocator-support/agnocast_heaphook/src/lib.rs#L238-L248

https://github.com/tier4/agnocast/blob/heaphook-custom-allocator-support/agnocast_heaphook/src/lib.rs#L315-L323

r1ru commented on agnocast_heaphook/src/lib.rs:
```diff
@@ -524,22 +449,21 @@ fn init_tlsf() {
-    tlsf.insert_free_block(pool);
-
-    TLSF.set(Mutex::new(tlsf));
+    assert!(TLSF.set(TLSFAllocator::new(pool)).is_ok())
```
fixed: https://github.com/tier4/agnocast/pull/686/commits/079cb6dd0f5e4d9a1ce6a486b5696cf4f7131451

r1ru commented on agnocast_heaphook/src/lib.rs:
```diff
@@ -138,12 +137,14 @@ extern "C" fn post_fork_handler_in_child() {
-type TlsfType = Tlsf<'static, FLBitmap, SLBitmap, FLLEN, SLLEN>;
-static TLSF: OnceLock<Mutex<TlsfType>> = OnceLock::new();
+static TLSF: OnceLock<TLSFAllocator> = OnceLock::new();
+
```
fixed: https://github.com/tier4/agnocast/pull/686/commits/67c39597d2efcca03d216f1749d49fe373a8313b

atsushi421 commented:
@r1ru 
Please fix [this](https://github.com/tier4/agnocast/actions/runs/17482528358/job/49655623960?pr=686).

r1ru commented:
@atsushi421 
fixed: https://github.com/tier4/agnocast/pull/686/commits/58c2a006b22638506d49cbb66efc4b3d6ce7a79f
Please check again.
